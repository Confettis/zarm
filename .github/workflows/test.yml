name: test
on: [push, pull_request]

jobs:
  setup:
    runs-on: ubuntu-latest
    steps:
      - name: checkout
        uses: actions/checkout@master

      - name: cache yarn.lock
        uses: actions/cache@v2
        with:
          path: package-temp-dir
          key: lock-${{ github.sha }}

      - name: hack for singe file
        run: |
          if [ ! -d "package-temp-dir" ]; then
            mkdir package-temp-dir
          fi
          cp yarn.lock package-temp-dir

      - name: cache node_modules
        id: node_modules_cache_id
        uses: actions/cache@v2
        with:
          path: node_modules
          key: node_modules-${{ hashFiles('**/package-temp-dir/yarn.lock') }}

      - name: install
        if: steps.node_modules_cache_id.outputs.cache-hit != 'true'
        run: yarn

  setup-react-16:
    runs-on: ubuntu-latest
    env:
      REACT: 16
    steps:
      - name: checkout
        uses: actions/checkout@master

      - name: install react 16
        run: lerna exec --scope zarm -- yarn install-react-16
    needs: setup

  compile-cli:
    runs-on: ubuntu-latest
    steps:
      - name: checkout
        uses: actions/checkout@master

      - name: restore cache from yarn.lock
        uses: actions/cache@v2
        with:
          path: package-temp-dir
          key: lock-${{ github.sha }}

      - name: restore cache from node_modules
        uses: actions/cache@v2
        with:
          path: node_modules
          key: node_modules-${{ hashFiles('**/package-temp-dir/yarn.lock') }}

      - name: cache zarm-cli/lib
        uses: actions/cache@v2
        with:
          path: packages/zarm-cli/lib
          key: cli-lib-${{ github.sha }}

      - name: compile cli
        run: lerna exec --scope @zarm-design/cli -- yarn build
    needs: setup

  compile:
    runs-on: ubuntu-latest
    steps:
      - name: checkout
        uses: actions/checkout@master

      - name: compile zarm
        run: lerna exec --scope zarm -- yarn build:es && yarn build:lib
    needs: compile-cli

  compile-react-16:
    runs-on: ubuntu-latest
    steps:
      - name: checkout
        uses: actions/checkout@master

      - name: compile
        run: lerna exec --scope zarm -- yarn build:es && yarn build:lib
    needs: [setup-react-16, compile-cli]

  lint:
    runs-on: ubuntu-latest
    steps:
      - name: checkout
        uses: actions/checkout@master

      - name: restore cache from yarn.lock
        uses: actions/cache@v2
        with:
          path: package-temp-dir
          key: lock-${{ github.sha }}

      - name: restore cache from node_modules
        uses: actions/cache@v2
        with:
          path: node_modules
          key: node_modules-${{ hashFiles('**/package-temp-dir/yarn.lock') }}

      - name: lint
        run: yarn lint
    needs: setup

  react-17:
    name: react@17.x
    runs-on: ubuntu-latest
    steps:
      - name: checkout
        uses: actions/checkout@master
        with:
          # codecov.io requires a fetch depth > 1.
          fetch-depth: 2

      - name: test
        run: lerna exec --scope zarm -- yarn coverage
      
      - name: coverage
        run: bash <(curl -s https://codecov.io/bash)
    needs: compile-cli

  react-17-dist:
    name: react@17.x / dist
    runs-on: ubuntu-latest
    steps:
      - name: checkout
        uses: actions/checkout@master

      - name: dist
        run: lerna exec --scope zarm -- yarn build:umd
        env:
          NODE_OPTIONS: --max_old_space_size=4096

      - name: check
        run: lerna exec --scope zarm -- node ./tests/dekko/dist.test.js

      - name: bundlesize
        run: lerna exec --scope zarm -- yarn bundlesize
        env:
          BUNDLESIZE_GITHUB_TOKEN: ${{ secrets.BUNDLESIZE_GITHUB_TOKEN }}
    needs: compile-cli

  react-16:
    name: react@16.x
    runs-on: ubuntu-latest
    env:
      REACT: 16
    steps:
      - name: checkout
        uses: actions/checkout@master

      - name: install react 16
        run: lerna exec --scope zarm -- yarn install-react-16

      - name: test
        run: lerna exec --scope zarm -- yarn test
    needs: [setup-react-16, compile-cli]

  react-16-dist:
    name: react@16.x / dist
    runs-on: ubuntu-latest
    env:
      REACT: 16
    steps:
      - name: checkout
        uses: actions/checkout@master

      - name: install react 16
        run: lerna exec --scope zarm -- yarn install-react-16

      - name: dist
        run: lerna exec --scope zarm -- yarn build:umd
        env:
          NODE_OPTIONS: --max_old_space_size=4096

      - name: check
        run: lerna exec --scope zarm -- node ./tests/dekko/dist.test.js

      - name: bundlesize
        run: lerna exec --scope zarm -- yarn bundlesize
        env:
          BUNDLESIZE_GITHUB_TOKEN: ${{ secrets.BUNDLESIZE_GITHUB_TOKEN }}
    needs: [setup-react-16, compile-cli]
