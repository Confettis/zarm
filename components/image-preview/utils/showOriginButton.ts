import { Images } from '../PropsType';

const showOriginButton = (images: Images, activeIndex: number | undefined): boolean => {
  if (
    images &&
    typeof activeIndex === 'number' &&
    images[activeIndex] &&
    images[activeIndex].originUrl
    // && !images[activeIndex].loaded
  ) {
    return true;
  }
  return false;
};

export default showOriginButton;
