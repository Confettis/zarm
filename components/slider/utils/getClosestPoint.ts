import type PropsType from '../PropsType';

export default function getClosestPoint(
  val: number,
  { marks, step, min, max }: Pick<PropsType, 'marks' | 'step' | 'min' | 'max'>,
) {
  const points = Object.keys(marks || {}).map(parseFloat);
  if (step !== null) {
    const maxSteps = Math.floor((max - min) / step);
    const steps = Math.min((val - min) / step, maxSteps);
    const closestStep = Math.round(steps) * step + min;
    points.push(closestStep);
  }
  const diffs = points.map((point) => Math.abs(val - point));

  return points[diffs.indexOf(Math.min(...diffs))];
}
