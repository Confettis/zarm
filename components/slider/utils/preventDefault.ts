import type { MouseEvent } from 'react';

export default function preventDefault(event: MouseEvent) {
  event.preventDefault();
}
