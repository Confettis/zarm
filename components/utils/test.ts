export function flushMicroTasks() {
  return new Promise((resolve) => setTimeout(resolve, 0));
}

export function noop() {}
