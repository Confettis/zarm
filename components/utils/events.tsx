import { noop } from './test';

let supportsPassive = false;
try {
  const opts = Object.defineProperty({}, 'passive', {
    get: () => {
      supportsPassive = true;
      return true;
    },
  });
  window.addEventListener('test', noop, opts);
  // eslint-disable-next-line no-empty
} catch (e) {}

declare global {
  interface Element {
    attachEvent(event: string, listener: EventListener): boolean;
    detachEvent(event: string, listener: EventListener): void;
  }
}

export default {
  supportsPassiveEvents: supportsPassive,
  on(
    el: Element,
    type: string,
    callback: EventListener,
    options: AddEventListenerOptions | boolean = { passive: false },
  ) {
    if (el.addEventListener) {
      el.addEventListener(type, callback, supportsPassive ? options : false);
    } else {
      el.attachEvent(`on${type}`, () => {
        callback.call(el);
      });
    }
  },

  off(
    el: Element,
    type: string,
    callback: EventListener,
    options: AddEventListenerOptions | boolean = { passive: false },
  ) {
    if (el.removeEventListener) {
      el.removeEventListener(type, callback, supportsPassive ? options : false);
    } else {
      el.detachEvent(`on${type}`, callback);
    }
  },

  once(
    el: Element,
    type: string,
    callback: EventListener,
    options: AddEventListenerOptions | boolean = { passive: false },
  ) {
    const typeArray = type.split(' ');
    const recursiveFunction = (e: Event) => {
      if (e.target) {
        e.target.removeEventListener(e.type, recursiveFunction, supportsPassive ? options : false);
      }
      return callback(e);
    };

    for (let i = typeArray.length - 1; i >= 0; i -= 1) {
      this.on(el, typeArray[i], recursiveFunction);
    }
  },
};
