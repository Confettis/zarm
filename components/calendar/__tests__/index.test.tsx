import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Calendar from '../index';
import { date7, date2, date6 } from '../../utils/test';

const prefixCls = 'za-calendar';

describe('Calendar', () => {
  it('calendar render', () => {
    const wrapper = mount(<Calendar defaultValue={date2} min={date2} max={date6} multiple={false} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find(`.${prefixCls}__day--selected`)).toHaveLength(1);
  });

  it('should trigger onChange when single click', async () => {
    const onChangeFn = jest.fn();
    const wrapper = mount(<Calendar min={date7} multiple={false} onChange={onChangeFn} />);

    const day2 = wrapper.find(`.${prefixCls}__day`).at(15);
    expect(day2.childAt(0).text()).toBe('16');
    day2.simulate('click');
    expect(onChangeFn).toBeCalled();
  });

  it('should trigger onChange when double click', () => {
    const onChangeFn = jest.fn();
    const wrapper = mount(<Calendar multiple={false} onChange={onChangeFn} />);

    const day3 = wrapper.find(`.${prefixCls}__day`).at(30);
    const day4 = wrapper.find(`.${prefixCls}__day`).at(20);
    day3.simulate('click');
    day4.simulate('click');
    expect(onChangeFn).toBeCalled();
  });
});
