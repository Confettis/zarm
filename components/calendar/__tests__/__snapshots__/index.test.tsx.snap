// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar calendar render 1`] = `
<CalendarView
  dateRender={[Function]}
  defaultValue={2017-11-03T07:00:00.000Z}
  disabledDate={[Function]}
  max={2017-11-06T07:00:00.000Z}
  min={2017-11-03T07:00:00.000Z}
  multiple={false}
  prefixCls="za-calendar"
>
  <div
    className="za-calendar"
  >
    <ul
      className="za-calendar__bar"
    >
      <li
        className="za-calendar__bar__item"
        key="日"
      >
        日
      </li>
      <li
        className="za-calendar__bar__item"
        key="一"
      >
        一
      </li>
      <li
        className="za-calendar__bar__item"
        key="二"
      >
        二
      </li>
      <li
        className="za-calendar__bar__item"
        key="三"
      >
        三
      </li>
      <li
        className="za-calendar__bar__item"
        key="四"
      >
        四
      </li>
      <li
        className="za-calendar__bar__item"
        key="五"
      >
        五
      </li>
      <li
        className="za-calendar__bar__item"
        key="六"
      >
        六
      </li>
    </ul>
    <section
      className="za-calendar__body"
    >
      <CalendarMonthView
        dateMonth={2017-10-31T19:00:00.000Z}
        dateRender={[Function]}
        disabledDate={[Function]}
        key="2017-10"
        max={2017-11-05T19:00:00.000Z}
        min={2017-11-02T19:00:00.000Z}
        onDateClick={[Function]}
        prefixCls="za-calendar"
        value={
          Array [
            2017-11-02T19:00:00.000Z,
          ]
        }
      >
        <section
          className="za-calendar__month"
          key="2017-10"
          title="2017年11月"
        >
          <ul>
            <li
              className="za-calendar__day za-calendar__day--disabled firstday-3"
              key="2017-10-1"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                1
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-2"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                2
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--selected"
              key="2017-10-3"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                3
              </div>
            </li>
            <li
              className="za-calendar__day d6"
              key="2017-10-4"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                4
              </div>
            </li>
            <li
              className="za-calendar__day d7"
              key="2017-10-5"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                5
              </div>
            </li>
            <li
              className="za-calendar__day"
              key="2017-10-6"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                6
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-7"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                7
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-8"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                8
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-9"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                9
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-10"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                10
              </div>
            </li>
            <li
              className="za-calendar__day d6 za-calendar__day--disabled"
              key="2017-10-11"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                11
              </div>
            </li>
            <li
              className="za-calendar__day d7 za-calendar__day--disabled"
              key="2017-10-12"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                12
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-13"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                13
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-14"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                14
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-15"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                15
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-16"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                16
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-17"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                17
              </div>
            </li>
            <li
              className="za-calendar__day d6 za-calendar__day--disabled"
              key="2017-10-18"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                18
              </div>
            </li>
            <li
              className="za-calendar__day d7 za-calendar__day--disabled"
              key="2017-10-19"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                19
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-20"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                20
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-21"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                21
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-22"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                22
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-23"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                23
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-24"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                24
              </div>
            </li>
            <li
              className="za-calendar__day d6 za-calendar__day--disabled"
              key="2017-10-25"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                25
              </div>
            </li>
            <li
              className="za-calendar__day d7 za-calendar__day--disabled"
              key="2017-10-26"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                26
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-27"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                27
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-28"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                28
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-29"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                29
              </div>
            </li>
            <li
              className="za-calendar__day za-calendar__day--disabled"
              key="2017-10-30"
              onClick={[Function]}
            >
              <div
                className="za-calendar__day__content"
              >
                30
              </div>
            </li>
          </ul>
        </section>
      </CalendarMonthView>
    </section>
  </div>
</CalendarView>
`;
